### PR TITLE
Make broadcast settings visible on users of with_container_arithmetic

### PR DESCRIPTION
**Pull request by inducer**
This is to support use cases like https://github.com/illinois-ceesd/mirgecom/pull/331, where the information that `with_container_arithmetic` has about the desired broadcast behavior of a type would also be useful to have. To make it available, this just shoves it into some (undocumented, for now) class attributes. In the linked use case, this would help (e.g.) avoid the need for hardcoding `DOFArray`. Instead, it could check whether `np.ndarray` is in `type(obj)._outer_bcast_names` and then follow that behavior.

I also have a distinct feeling that this might also be a bad idea. @alexfikl, got a gut feeling on this one?

cc @majosm

### CODE REVIEW

**Comment by alexfikl** [comment]
Is it worth moving the `outer` function somewhere in `arraycontext.container`? That way it would keep all of this a bit more contained if we ever decide to change it.

Besides that, no particular feeling of impending doom on my side :shrug: 

(There's also a stray `print` left in there)

File: arraycontext/container/arithmetic.py [hunk 1]
```diff
@@ -390,6 +394,7 @@ def {fname}(arg2, arg1):
 
         # }}}
 
+        print(gen.get())
```
Comment (alexfikl) [inline arraycontext/container/arithmetic.py:397]:
```suggestion
```

**Review by alexfikl — APPROVED** [review]

**Comment by inducer** [comment]
> Is it worth moving the `outer` function somewhere in `arraycontext.container`?

I'd be all for that. @majosm? (Added you to the repo just in case.)

**Comment by inducer** [comment]
Thanks for taking a look!

File: arraycontext/container/arithmetic.py [hunk 2]
```diff
@@ -280,6 +280,10 @@ def tup_str(t: Tuple[str, ...]) -> str:
             else:
                 return "(%s,)" % ", ".join(t)
 
+        gen(f"cls._outer_bcast_names = {tup_str(outer_bcast_type_names)}")
```
Comment (inducer) [inline arraycontext/container/arithmetic.py:283]:
```suggestion
        gen(f"cls._outer_bcast_types = {tup_str(outer_bcast_type_names)}")
```

**Comment by majosm** [comment]
> > Is it worth moving the `outer` function somewhere in `arraycontext.container`?
> 
> I'd be all for that, especially since it'd help avoid use of private attributes across package boundaries. @majosm? (Added you to the repo just in case.)

Just wondering: how come the specifications in `with_container_arithmetic` don’t become public? Would it be useful for external code to have access to that so it can build on top of the existing mapping functionality?

**Comment by inducer** [comment]
> Just wondering: how come the specifications in `with_container_arithmetic` don’t become public? Would it be useful for external code to have access to that so it can build on top of the existing mapping functionality?

I could very well see it being useful. I just have commitment anxiety about the interface... it seems kind of janky and poorly defined.

**Comment by majosm** [comment]
@inducer Where should I stick `outer`?

**Comment by inducer** [comment]
`container.traversal`, and importable from the root?

**Comment by majosm** [comment]
#46

**Comment by inducer** [comment]
Closing in favor of #46.

**Pull request closed without merging**